### PR TITLE
External VM

### PR DESCRIPTION
**Pull request by LukasGasior1**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -126,7 +127,7 @@ val root = project.in(file("."))
         warnOnUnverifiedFiles = false,
         warnOnUnusedVerifications = false
       ),
-      dist in Universal := ((dist in Universal) dependsOn verify).value
```
Comment (LukasGasior1) [inline build.sbt]: TODO: bring it back

File: build.sbt [hunk 2]
```diff
@@ -140,8 +141,7 @@ scalacOptions := Seq(
   "-feature",
   "-Xfatal-warnings",
   "-Xlint:unsound-match",
-  "-Ywarn-inaccessible",
-  "-Ywarn-unused-import"
```
Comment (LukasGasior1) [inline build.sbt]: TODO: bring it back

File: src/ets/scala/io/iohk/ethereum/ets/blockchain/ScenarioSetup.scala [hunk 3]
```diff
@@ -1,30 +1,41 @@
...
+import io.iohk.ethereum.utils.{BlockchainConfig, Config}
 import io.iohk.ethereum.vm.VM
 import org.spongycastle.util.encoders.Hex
 
 import scala.util.{Failure, Success, Try}
 
+object ScenarioSetup {
+  implicit lazy val actorSystem = ActorSystem("mantis_system")
+  lazy val extvm = new ExtVMInterface("127.0.0.1", 8888, BlockchainConfig(Config.config))
+  val vm = VM
```
Comment (LukasGasior1) [inline src/ets/scala/io/iohk/ethereum/ets/blockchain/ScenarioSetup.scala]: TODO: add parameter for tests to run in local/remote vm (?)

File: src/main/scala/io/iohk/ethereum/extvm/VMServer.scala [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    new VMServer(in, out).run()
+  }
+}
+
+class VMServer(
+    override val in: SinkQueueWithCancel[ByteString],
+    override val out: SourceQueueWithComplete[ByteString])
+  extends MessageUtils with Logger {
+
+  class StorageCache {
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/extvm/VMServer.scala]: TODO: extract all the inner classes into top level classes

File: src/main/scala/io/iohk/ethereum/vm/ProgramContext.scala [hunk 5]
```diff
@@ -25,10 +27,52 @@ object ProgramContext {
...
+    receivingAddr: Address,
+    startGas: BigInt,
+    world: W,
+    blockchainConfig: BlockchainConfig
+  ): ProgramContext[W, S] = {
+    val evmConfig = EvmConfig.forBlock(env.blockHeader.number, blockchainConfig)
+    ProgramContext(env, receivingAddr, startGas, world, Some(blockchainConfig), evmConfig)
+  }
+
+  def apply[W <: WorldStateProxy[W, S], S <: Storage[S]](
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/ProgramContext.scala]: TODO: do we really need that many `apply`s. Look for a smarter solution

File: src/main/protobuf/extvm/msg.proto [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  bytes ownerAddr = 1;
+  bytes callerAddr = 2;
+  bytes originAddr = 3;
+  bytes contractCode = 4;
+  bytes inputData = 5;
+  bytes callValue = 6;
+  bytes gasPrice = 7;
+  bytes gasProvided = 8;
+  int32 callDepth = 9;
+  bytes receivingAddr = 10;
```
Comment (rtkaczyk) [inline src/main/protobuf/extvm/msg.proto]: TODO: `receivingAddr` is redundant

File: src/main/protobuf/extvm/msg.proto [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  bytes number = 9;
+  bytes gasLimit = 10;
+  bytes gasUsed = 11;
+  int64 unixTimestamp = 12;
+  bytes extraData = 13;
+  bytes mixHash = 14;
+  bytes nonce = 15;
+}
+
+message BlockchainConfig {
```
Comment (rtkaczyk) [inline src/main/protobuf/extvm/msg.proto]: TODO: something... probably... ¯\_(ツ)_/¯

File: src/main/protobuf/extvm/msg.proto [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+message GetAccount {
+  bytes address = 1;
+}
+
+message Account {
+  bytes nonce = 1;
+  bytes balance = 2;
+  //FIXME: the hashes required because of an implementation detail, should be gone in the final solution
+  bytes storageHash = 3;
+  bytes codeHash = 4;
```
Comment (rtkaczyk) [inline src/main/protobuf/extvm/msg.proto]: TODO: change `codeHash` to a boolean flag like: `nonEmptyCode`; `storageHash` shouldn't be required

**Comment by loverdos** [comment]
@LukasGasior1 @rtkaczyk Why not add your `TODO` comments in the code? If we merge this and the PR is closed, we never look back and they get forgotten.

**Comment by loverdos** [comment]
When compiling, I get:

```
[warn] There may be incompatibilities among your library dependencies.
[warn] Here are some of the libraries that were evicted:
[warn] 	* com.typesafe.akka:akka-stream_2.12:2.4.17 -> 2.5.8
[warn] 	* com.typesafe.akka:akka-actor_2.12:2.4.17 -> 2.5.8
[warn] Run 'evicted' to see detailed eviction warnings
```
and then `sbt evicted` gives:

```
[warn] There may be incompatibilities among your library dependencies.
[warn] Here are some of the libraries that were evicted:
[warn] 	* com.typesafe.akka:akka-stream_2.12:2.4.17 -> 2.5.8 (caller: com.typesafe.akka:akka-http-core_2.12:10.0.6, mantis:mantis_2.12:1.0-daedalus-rc1)
[warn] 	* com.typesafe.akka:akka-actor_2.12:2.4.17 -> 2.5.8 (caller: mantis:mantis_2.12:1.0-daedalus-rc1, com.typesafe.akka:akka-stream_2.12:2.5.8 (), com.typesafe.akka:akka-agent_2.12:2.4.17 (), com.typesafe.akka:akka-testkit_2.12:2.4.17 (), com.typesafe.akka:akka-slf4j_2.12:2.4.17 (), com.typesafe.akka:akka-parsing_2.12:10.0.6)
[info] Here are other libraries that were evicted:
[info] 	* com.typesafe:config:(1.3.0, 1.2.0) -> 1.3.2 (caller: com.typesafe.akka:akka-actor_2.12:2.5.8 (), com.typesafe.akka:akka-actor_2.12:2.4.17 (), com.typesafe:ssl-config-core_2.12:0.2.2)
[info] 	* org.scala-lang.modules:scala-parser-combinators_2.12:1.0.4 -> 1.0.5 (caller: mantis:mantis_2.12:1.0-daedalus-rc1, com.typesafe:ssl-config-core_2.12:0.2.2)
[info] 	* org.slf4j:slf4j-api:(1.7.21, 1.7.16, 1.7.22) -> 1.7.23 (caller: org.scorexfoundation:iodb_2.12:0.3.0, org.consensusresearch:scrypto_2.12:1.2.0-RC3, com.typesafe.akka:akka-slf4j_2.12:2.4.17 (), ch.qos.logback:logback-classic:1.1.9)
[info] 	* org.json4s:json4s-core_2.12:3.5.0 -> 3.5.1 (caller: org.json4s:json4s-native_2.12:3.5.1, de.heikoseeberger:akka-http-json4s_2.12:1.11.0)
[info] 	* com.typesafe.akka:akka-http_2.12:(10.0.0, 10.0.5) -> 10.0.6 (caller: ch.megard:akka-http-cors_2.12:0.2.1, de.heikoseeberger:akka-http-json4s_2.12:1.11.0, mantis:mantis_2.12:1.0-daedalus-rc1)
```

**Comment by rtkaczyk** [comment]
> @LukasGasior1 @rtkaczyk Why not add your TODO comments in the code? If we merge this and the PR is closed, we never look back and they get forgotten.

Because we want to close this ASAP, as we're both responsible for this code, which is not the best way to cooperate. The comments won't be lost when the PR is closed, and we're already working on them.

Once this gets merged the regular development flow will resume.

**Comment by rtkaczyk** [comment]
>[warn] 	* com.typesafe.akka:akka-stream_2.12:2.4.17 -> 2.5.8
[warn] 	* com.typesafe.akka:akka-actor_2.12:2.4.17 -> 2.5.8

Good catch, we'll take a look at that in a follow-up PR.

**Pull request merged**